llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp

```diff
@@ -4652,21 +4652,25 @@ SDValue TargetLowering::SimplifySetCC(EVT VT, SDValue N0, SDValue N1,
                           DAG.getConstant(C1 & Imm, dl, ExtDstTy), Cond);
     } else if ((N1C->isZero() || N1C->isOne()) &&
                (Cond == ISD::SETEQ || Cond == ISD::SETNE)) {
-      // SETCC (SETCC), [0|1], [EQ|NE]  -> SETCC
-      if (N0.getOpcode() == ISD::SETCC &&
+      // SETCC (X), [0|1], [EQ|NE]  -> X if X is known 0/1
+      if ((N0.getOpcode() == ISD::SETCC || VT != MVT::i1) &&
```

Comment:
I don't quite understand the `VT != MVT::i1` here. Doesn't that break things like `setcc (setcc ...))`

Reply:
It will do the transform if it is a setcc, or if the type _isn't_ an i1. i1 types are trivially true for the known bits combine, and handled further below when the foldBools flag is set.